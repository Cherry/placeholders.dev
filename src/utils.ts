import PoPs from '@adaptivelink/pops';

/* security headers */
export const addHeaders = {
	'X-XSS-Protection': '1; mode=block',
	'X-Frame-Options': 'DENY',
	'X-Content-Type-Options': 'nosniff',
	'Referrer-Policy': 'no-referrer-when-downgrade',
	'Strict-Transport-Security': 'max-age=31536000; includeSubDomains; preload',
	'Feature-Policy': [
		'geolocation \'none\';',
		'midi \'none\';',
		'sync-xhr \'none\';',
		'microphone \'none\';',
		'camera \'none\';',
		'magnetometer \'none\';',
		'gyroscope \'none\';',
		'speaker \'none\';',
		'fullscreen \'none\';',
		'payment \'none\';',
	].join(' '),
	'Content-Security-Policy': [
		'default-src \'self\';',
		'script-src \'self\' cdnjs.cloudflare.com static.cloudflareinsights.com;',
		'style-src \'self\' cdnjs.cloudflare.com \'unsafe-inline\' fonts.googleapis.com;',
		'img-src \'self\' data: images.placeholders.dev;',
		'child-src \'none\';',
		'font-src \'self\' fonts.gstatic.com cdnjs.cloudflare.com;',
		'connect-src \'self\';',
		'prefetch-src \'none\';',
		'object-src \'none\';',
		'form-action \'none\';',
		'frame-ancestors \'none\';',
		'upgrade-insecure-requests;',
	].join(' '),
} as const;

/* caching */
export const cacheTtl = 60 * 60 * 24 * 90; // 90 days
export const filesRegex = /(.*\.(ac3|avi|bmp|br|bz2|css|cue|dat|doc|docx|dts|eot|exe|flv|gif|gz|htm|html|ico|img|iso|jpeg|jpg|js|json|map|mkv|mp3|mp4|mpeg|mpg|ogg|pdf|png|ppt|pptx|qt|rar|rm|svg|swf|tar|tgz|ttf|txt|wav|webp|webm|webmanifest|woff|woff2|xls|xlsx|xml|zip))$/;

/* HTML rewriter for more accurate count of Cloudflare PoPs */
export const CloudflarePoPs = PoPs.cloudflare.code.length;
export const description = `Generate super fast placeholder images powered by Cloudflare Workers in ${CloudflarePoPs}+ edge locations.`;
export class PoPsRewriter implements HTMLRewriterElementContentHandlers {
	element(element: Element) {
		if (element.tagName === 'title') {
			element.setInnerContent(description);
		} else if (element.tagName === 'meta' && (element.getAttribute('name') === 'description' || element.getAttribute('name') === 'twitter:description' || element.getAttribute('property') === 'og:description')) {
			element.setAttribute('content', description);
		} else if (element.tagName === 'span' && element.hasAttribute('class') && element.getAttribute('class')?.includes('edgeLocations')) {
			element.setInnerContent(String(CloudflarePoPs));
		}
	}
}

export const availableImageOptions = [
	'width',
	'height',
	'text',
	'dy',
	'fontFamily',
	'fontWeight',
	'fontSize',
	'bgColor',
	'textColor',
	'darkBgColor',
	'darkTextColor',
	'textWrap',
] as const;

export function getKeys<T extends object>(obj: T) {
	return Object.keys(obj) as Array<keyof T>;
}
