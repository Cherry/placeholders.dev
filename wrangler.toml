name = "placeholders-dev"
workers_dev = true
compatibility_date = "2023-12-09"
compatibility_flags = ["nodejs_compat"]
main = "./src/index.ts"

[site]
bucket = "./public"

[env.staging]
name = "placeholders-staging"
workers_dev = true

[env.production]
name = "placeholders"
workers_dev = false
routes = [
	{pattern = "placeholders.dev/*", zone_id = "51bfc6eeecc90565bd116ee3ebb5e3a8"},
	{pattern = "images.placeholders.dev/*", zone_id = "51bfc6eeecc90565bd116ee3ebb5e3a8"}
]
logpush = true


analytics_engine_datasets = [
	{ binding = "PLACEHOLDERS_ANALYTICS" }
]
