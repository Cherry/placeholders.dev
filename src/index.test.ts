import {
	afterAll,
	beforeAll,
	describe,
	expect,
	it,
	test,
} from 'vitest';
import { unstable_dev } from 'wrangler';

import { getKeys } from './utils';

import type { UnstableDevWorker } from 'wrangler';

describe('Worker', () => {
	let worker: UnstableDevWorker;

	beforeAll(async () => {
		worker = await unstable_dev('src/index.ts', {
			experimental: {
				disableExperimentalWarning: true,
			},
		});
	});

	afterAll(async () => {
		await worker.stop();
	});

	it('should return html landing page', async () => {
		const req = new Request('https://example.com', { method: 'GET' });
		const resp = await worker.fetch(req.url);
		expect(resp.status).toBe(200);

		// check if html is returned
		const headers = resp.headers;
		expect(headers.get('content-type')).toBe('text/html; charset=utf-8');
		const text = await resp.text();
		expect(text).toMatch(/^<!DOCTYPE html>/);
	});

	it('should sanitize for XSS', async () => {
		const req = new Request('https://example.com/api/?width=350&height=100&text=Hello%20World&bgColor=%22%3E%3Cscript%3Ealert(%22XSS%22);%3C/script%3E', { method: 'GET' });
		const resp = await worker.fetch(req.url);
		expect(resp.status).toBe(200);

		const text = await resp.text();
		expect(text).toMatchSnapshot();
	});

	it('should sanitize for CSS prop injection', async () => {
		const req = new Request('https://example.com/api/?width=450&height=450&text=James&fontFamily=test;background:url(https://avatars.githubusercontent.com/u/856748?v=4)&textWrap=true', { method: 'GET' });
		const resp = await worker.fetch(req.url);
		expect(resp.status).toBe(200);

		const text = await resp.text();
		expect(text).toMatchSnapshot();
	});

	test.each([
		// basic tests
		[
			{
				width: 350,
				height: 100,
			},
		],
		[
			{
				width: 200,
				height: 100,
				bgColor: '#000',
				textColor: 'rgba(255,255,255,0.5)',
			},
		],
		[
			{
				width: 140,
				height: 100,
				bgColor: '#313131',
				textColor: '#dfdfde',
			},
		],
		[
			{
				width: 350,
				height: 100,
				text: 'placeholders.dev',
			},
		],
		[
			{
				width: 1055,
				height: 100,
				text: 'Hello World',
				bgColor: '#434343',
				textColor: '#dfdfde',
			},
		],
		// text wrapping
		[
			{
				width: 250,
				height: 200,
				text: 'This text is too long',
				bgColor: '#f7f6f6',
				textWrap: false,
			},
		],
		[
			{
				width: 250,
				height: 200,
				text: 'This text is too long',
				bgColor: '#f7f6f6',
				textWrap: true,
			},
		],
	])('should return accurate svg image with query params %s', async (query) => {
		const searchParams = new URLSearchParams();
		for (const key of getKeys(query)) {
			searchParams.set(key, String(query[key]));
		}
		const req = new Request(`https://example.com/api/?${searchParams.toString()}`, { method: 'GET' });
		const resp = await worker.fetch(req.url);
		expect(resp.status).toBe(200);

		const text = await resp.text();
		expect(text).toMatchSnapshot();
	});
});
