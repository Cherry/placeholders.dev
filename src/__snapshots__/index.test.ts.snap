// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Worker > should return accurate svg image with query params { width: 140, height: 100, bgColor: '#313131', textColor: '#dfdfde' } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="140" height="100" viewBox="0 0 140 100"><rect fill="#313131" width="140" height="100"/><text fill="#dfdfde" font-family="sans-serif" font-size="20" dy="7" font-weight="bold" x="50%" y="50%" text-anchor="middle">140×100</text></svg>"`;

exports[`Worker > should return accurate svg image with query params { width: 200, height: 100, bgColor: '#000', textColor: 'rgba(255,255,255,0.5)' } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="200" height="100" viewBox="0 0 200 100"><rect fill="#000" width="200" height="100"/><text fill="rgba(255,255,255,0.5)" font-family="sans-serif" font-size="20" dy="7" font-weight="bold" x="50%" y="50%" text-anchor="middle">200×100</text></svg>"`;

exports[`Worker > should return accurate svg image with query params { width: 250, height: 200, text: 'This text is too long', bgColor: '#f7f6f6', textWrap: false } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="250" height="200" viewBox="0 0 250 200"><rect fill="#f7f6f6" width="250" height="200"/><text fill="rgba(0,0,0,0.5)" font-family="sans-serif" font-size="40" dy="14" font-weight="bold" x="50%" y="50%" text-anchor="middle">This text is too long</text></svg>"`;

exports[`Worker > should return accurate svg image with query params { width: 250, height: 200, text: 'This text is too long', bgColor: '#f7f6f6', textWrap: true } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="250" height="200" viewBox="0 0 250 200"><rect fill="#f7f6f6" width="250" height="200"/><foreignObject width="250" height="200"><div xmlns="http://www.w3.org/1999/xhtml" style="align-items: center;box-sizing: border-box;color: rgba(0,0,0,0.5);display: flex;font-family: sans-serif;font-size: 40px;font-weight: bold;height: 100%;line-height: 1.2;justify-content: center;padding: 0.5em;text-align: center;width: 100%;">This text is too long</div> </foreignObject></svg>"`;

exports[`Worker > should return accurate svg image with query params { width: 350, height: 100 } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="350" height="100" viewBox="0 0 350 100"><rect fill="#ddd" width="350" height="100"/><text fill="rgba(0,0,0,0.5)" font-family="sans-serif" font-size="20" dy="7" font-weight="bold" x="50%" y="50%" text-anchor="middle">350×100</text></svg>"`;

exports[`Worker > should return accurate svg image with query params { width: 350, height: 100, text: 'placeholders.dev' } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="350" height="100" viewBox="0 0 350 100"><rect fill="#ddd" width="350" height="100"/><text fill="rgba(0,0,0,0.5)" font-family="sans-serif" font-size="20" dy="7" font-weight="bold" x="50%" y="50%" text-anchor="middle">placeholders.dev</text></svg>"`;

exports[`Worker > should return accurate svg image with query params { width: 1055, height: 100, text: 'Hello World', bgColor: '#434343', textColor: '#dfdfde' } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="1055" height="100" viewBox="0 0 1055 100"><rect fill="#434343" width="1055" height="100"/><text fill="#dfdfde" font-family="sans-serif" font-size="20" dy="7" font-weight="bold" x="50%" y="50%" text-anchor="middle">Hello World</text></svg>"`;

exports[`Worker > should sanitize for CSS prop injection 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="450" height="450" viewBox="0 0 450 450"><rect fill="#ddd" width="450" height="450"/><foreignObject width="450" height="450"><div xmlns="http://www.w3.org/1999/xhtml" style="align-items: center;box-sizing: border-box;color: rgba(0,0,0,0.5);display: flex;font-family: testbackgroundurl(https//avatars.githubusercontent.com/u/856748?v=4);font-size: 90px;font-weight: bold;height: 100%;line-height: 1.2;justify-content: center;padding: 0.5em;text-align: center;width: 100%;">James</div> </foreignObject></svg>"`;

exports[`Worker > should sanitize for XSS 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="350" height="100" viewBox="0 0 350 100"><rect fill="#ddd" width="350" height="100"/><text fill="rgba(0,0,0,0.5)" font-family="sans-serif" font-size="20" dy="7" font-weight="bold" x="50%" y="50%" text-anchor="middle">Hello World</text></svg>"`;
